src/ir/context.rs

```diff
@@ -131,29 +141,46 @@ impl<'ctx> BindgenContext<'ctx> {
 
         let id = item.id();
         let is_type = item.kind().is_type();
+        let is_unnamed = is_type && item.expect_type().name().is_none();
         let old_item = self.items.insert(id, item);
         assert!(old_item.is_none(), "Inserted type twice?");
 
+        // Unnamed items can have an USR, but they can't be referenced from
+        // other sites explicitly and the USR can match if the unnamed items are
+        // nested, so don't bother tracking them.
         if is_type && declaration.is_some() {
-            let declaration = declaration.unwrap();
-            debug_assert_eq!(declaration, declaration.canonical());
-            if declaration.is_valid() {
-                let old = self.types.insert(declaration, id);
-                debug_assert_eq!(old, None);
-            } else if location.is_some() &&
-                      (location.unwrap().kind() == CXCursor_ClassTemplate ||
-                       location.unwrap().kind() == CXCursor_ClassTemplatePartialSpecialization) {
-                let old = self.types.insert(location.unwrap().canonical(), id);
-                debug_assert_eq!(old, None);
-            } else {
+            let mut declaration = declaration.unwrap();
+            if !declaration.is_valid() {
+                if let Some(location) = location {
+                    if location.kind() == CXCursor_ClassTemplate ||
+                       location.kind() == CXCursor_ClassTemplatePartialSpecialization {
+                        declaration = location;
+                    }
+                }
+            }
+            declaration = declaration.canonical();
+            if !declaration.is_valid() {
```

Comment:
Here you rebound `declaration` to the canonical one, which is invalid.

What does the `debug!` message end up saying, given the declaration you passed to it is the invalid one now?